luaui/Widgets/api_blueprint.lua

```diff
@@ -543,6 +567,381 @@ local BUILD_MODES_HANDLERS = {
+			totalBuildPower = totalBuildPower + groupData.power
+		end
+
+		if totalBuildPower > 0 then
+			local allBuildingsWithCost = {}
+			local totalBuildCost = 0
+			for _, building in ipairs(allBuildings) do
+				local unitDef = UnitDefs[building.unitDefID]
+				local cost = (unitDef and unitDef.cost) or 0
+				table.insert(allBuildingsWithCost, { building = building, cost = cost })
+				totalBuildCost = totalBuildCost + cost
+			end
+
+			-- 2. Partition the blueprint into cost-based linear chunks
+			local chunks = {}
+			local buildingIndex = 1
+			for i, groupData in ipairs(allBuilderGroups) do
+				local proportion = groupData.power / totalBuildPower
+				local targetCostForGroup = totalBuildCost * proportion
+				
+				local buildingsForGroup = {}
+				local accumulatedCost = 0
+				
+				if i == #allBuilderGroups then
+					-- Last group takes all remaining buildings
+					for j = buildingIndex, #allBuildingsWithCost do
+						table.insert(buildingsForGroup, allBuildingsWithCost[j].building)
+					end
+				else
+					while buildingIndex <= #allBuildingsWithCost do
+						local currentBuilding = allBuildingsWithCost[buildingIndex]
+						local costAfterAdding = accumulatedCost + currentBuilding.cost
+						
+						-- Stop if adding the next building makes the chunk's cost further from the target
+						if accumulatedCost > 0 and math.abs(costAfterAdding - targetCostForGroup) > math.abs(accumulatedCost - targetCostForGroup) then
+							break
+						end
+						
+						table.insert(buildingsForGroup, currentBuilding.building)
+						accumulatedCost = costAfterAdding
+						buildingIndex = buildingIndex + 1
+					end
+				end
+
+				table.insert(chunks, { groupData = groupData, buildings = buildingsForGroup })
+			end
+			
+			-- 3. Resolve chunks and create final orders
+			local finalOrders = {}
+			local leftovers = {}
+			for _, groupData in ipairs(allBuilderGroups) do
+				finalOrders[groupData.key] = {}
+			end
+			
+			for _, chunk in ipairs(chunks) do
+				local groupData = chunk.groupData
+				local builderGroup = groupData.group
+				local groupSide = groupData.side
+				local groupKey = groupData.key
+
+				for _, building in ipairs(chunk.buildings) do
+					if canBuild(builderGroup, building, groupSide, true) then
+						local substitutedUnitName = SubLogic.getEquivalentUnitDefID(building.originalName, groupSide)
+						if substitutedUnitName then
+							local substitutedUnitDefID = UnitDefNames[substitutedUnitName] and UnitDefNames[substitutedUnitName].id
+							if substitutedUnitDefID then
+								table.insert(finalOrders[groupKey], { -substitutedUnitDefID, { building.position[1], building.position[2], building.position[3], building.facing }, newOpts })
+							end
+						end
+					else
+						table.insert(leftovers, building)
+					end
+				end
+			end
+			
+			-- 4. Handle leftovers
+			local leftoverTracker = {} -- category -> index
+			for _, building in ipairs(leftovers) do
```

Comment:
This is kind of a procedural catch-all I don't love. We could, with perfect knowledge of the system and good tagging of builders and buildings, plan our tasks so that this step is not necessary. It does work well however and the idea of making the planner smart enough to catch things like t1 air cons being able to build t2 air labs seems daunting.